src/tests/ftest/util/run_utils.py

```diff
@@ -137,23 +152,67 @@ def _process_task(self, task, command):
 
         # Populate the a list of unique output for each NodeSet
         for code in sorted(results):
-            output_data = list(task.iter_buffers(results[code]))
-            if not output_data:
-                output_data = [["<NONE>", results[code]]]
-            for output, output_hosts in output_data:
+            stdout_data = self._sanitize_iter_data(
+                results[code], list(task.iter_buffers(results[code])), "<NONE>")
+
+            for stdout_raw, stdout_hosts in stdout_data:
                 # In run_remote(), task.run() is executed with the stderr=False default.
                 # As a result task.iter_buffers() will return combined stdout and stderr.
-                stdout = []
-                for line in output.splitlines():
-                    if isinstance(line, bytes):
-                        stdout.append(line.decode("utf-8"))
-                    else:
-                        stdout.append(line)
-                self.output.append(
-                    self.ResultData(command, code, NodeSet.fromlist(output_hosts), stdout, False))
+                stdout = self._msg_tree_elem_to_list(stdout_raw)
+                stderr_data = self._sanitize_iter_data(
+                    stdout_hosts, list(task.iter_errors(stdout_hosts)), "")
+                for stderr_raw, stderr_hosts in stderr_data:
+                    stderr = self._msg_tree_elem_to_list(stderr_raw)
+                    self.output.append(
+                        ResultData(
+                            command, code, NodeSet.fromlist(stderr_hosts), stdout, stderr, False))
         if timed_out:
             self.output.append(
-                self.ResultData(command, 124, NodeSet.fromlist(timed_out), None, True))
+                ResultData(command, 124, NodeSet.fromlist(timed_out), None, None, True))
+
+    @staticmethod
+    def _sanitize_iter_data(hosts, data, default_entry):
+        """Ensure the data generated from an iter function has entries for each host.
+
+        Args:
+            hosts (list): lists of host which generated data
+            data (list): data from an iter function as a list
+            default_entry (object): entry to add to data for missing hosts in data
+
+        Returns:
+            list: a list of tuples of entries and list of hosts
+        """
+        if not data:
+            return [(default_entry, hosts)]
+
+        source_keys = NodeSet.fromlist(hosts)
+        data_keys = NodeSet()
+        for _, keys in data:
+            data_keys.add(NodeSet.fromlist(keys))
```

Comment:
FYI - I discovered this recently
```suggestion
            data_keys.updaten(keys)
```